drivers/wasapi/audio_driver_wasapi.cpp

```diff
@@ -376,87 +281,60 @@ Error AudioDriverWASAPI::audio_device_init(AudioDeviceWASAPI *p_device, bool p_i
+		streamflags |= AUDCLNT_STREAMFLAGS_AUTOCONVERTPCM | AUDCLNT_STREAMFLAGS_SRC_DEFAULT_QUALITY;
+		pwfex->nSamplesPerSec = mix_rate;
+		pwfex->nAvgBytesPerSec = pwfex->nSamplesPerSec * pwfex->nChannels * (pwfex->wBitsPerSample / 8);
+	}
 
-		// Due to WASAPI Shared Mode we have no control of the buffer size
-		if (!p_input) {
-			buffer_frames = max_frames;
+	if (!p_input) {
+		// Let WASAPI notify us when it wants more samples (push method).
+		streamflags |= AUDCLNT_STREAMFLAGS_EVENTCALLBACK;
+	}
 
-			int64_t latency = 0;
-			audio_output.audio_client->GetStreamLatency(&latency);
-			// WASAPI REFERENCE_TIME units are 100 nanoseconds per unit
-			// https://docs.microsoft.com/en-us/windows/win32/directshow/reference-time
-			// Convert REFTIME to seconds as godot uses for latency
-			real_latency = (float)latency / (float)REFTIMES_PER_SEC;
-		}
+	// We open in shared mode. Exclusive mode is more intricate and can result in lower latency (can even have DMA writes to external hardware)
+	// however as it is exclusive, other programs will not be able to play to the speakers anymore which is not suitable for most applications.
+	// This function will fail if the chosen output device is currently opened in another application in exclusive mode.
+	//
+	// Passing in 0 here for the buffer duration means that we will get the smallest possible buffer (also lowest latency).
+	hr = p_device->audio_client->Initialize(AUDCLNT_SHAREMODE_SHARED, streamflags, p_input ? REFTIMES_PER_SEC : 0, 0, pwfex, nullptr);
+	ERR_FAIL_COND_V_MSG(hr != S_OK, ERR_CANT_OPEN, "WASAPI: Initialize failed with error: 0x" + String::num_uint64(hr, 16) + ".");
 
-	} else {
-		IAudioClient3 *device_audio_client_3 = (IAudioClient3 *)p_device->audio_client;
-
-		// AUDCLNT_STREAMFLAGS_RATEADJUST is an invalid flag with IAudioClient3, therefore we have to use
-		// the closest supported mix rate supported by the audio driver.
-		mix_rate = pwfex->nSamplesPerSec;
-		print_verbose("WASAPI: mix_rate = " + itos(mix_rate));
-
-		UINT32 default_period_frames, fundamental_period_frames, min_period_frames, max_period_frames;
-		hr = device_audio_client_3->GetSharedModeEnginePeriod(
-				pwfex,
-				&default_period_frames,
-				&fundamental_period_frames,
-				&min_period_frames,
-				&max_period_frames);
-		if (hr != S_OK) {
-			print_verbose("WASAPI: GetSharedModeEnginePeriod failed with error 0x" + String::num_uint64(hr, 16) + ", falling back to IAudioClient.");
-			CoTaskMemFree(pwfex);
-			SAFE_RELEASE(output_device)
-			return audio_device_init(p_device, p_input, p_reinit, true);
-		}
+	if (!p_input) {
+		// This is the event that WASAPI will signal when it wants more data from us (push method).
+		// The initial state for this event must be unsignaled!
+		p_device->feed_event = CreateEventW(nullptr, FALSE, FALSE, nullptr);
+		ERR_FAIL_COND_V_MSG(p_device->feed_event == nullptr, ERR_CANT_OPEN, "WASAPI: Could not create event handle with error: " + String::num(GetLastError()) + ".");
```

Comment:
```suggestion
		ERR_FAIL_NULL_V_MSG(p_device->feed_event, ERR_CANT_OPEN, "WASAPI: Could not create event handle with error: " + String::num(GetLastError()) + ".");
```
This is correct, look up the checks if you don't trust me